### PR TITLE
Don't mark world state as stalled until a minimum time without progress is reached

### PR DESCRIPTION
**Pull request by ajsutton**
## PR description
Delays marking a world state download as stalled until at least 5 minutes has passed since the last progress was made. This is in addition to requiring a minimum number of requests without progress.

Additionally increases the minimum number of requests without progress from 100 to 1000.  Since restarting the world state download is expensive it's worth having high thresholds before giving up.

### CODE REVIEW

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java [hunk 1]
```diff
@@ -131,6 +136,18 @@ public void shouldResetRequestsSinceProgressCountWhenProgressIsMade() {
     assertThat(downloadState.getDownloadFuture()).isCompletedExceptionally();
   }
 
+  @Test
+  public void shouldNotBeStalledWhenMaxRequestsReachedButNotMinimumTime() {
```
Comment (mbaxter) [inline ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java]:
```suggestion
  public void shouldStallWhenMaxRequestsReachedAndTimeLimitExceeded() {
```

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java [hunk 2]
```diff
@@ -131,6 +136,18 @@ public void shouldResetRequestsSinceProgressCountWhenProgressIsMade() {
...
+    for (int i = 0; i < MAX_REQUESTS_WITHOUT_PROGRESS; i++) {
+      downloadState.requestComplete(false);
+      assertThat(downloadState.getDownloadFuture()).isNotDone();
+    }
+
+    clock.stepMillis(MIN_MILLIS_BEFORE_STALLING + 1);
+    downloadState.requestComplete(false);
+    assertThat(downloadState.getDownloadFuture()).isCompletedExceptionally();
```
Comment (mbaxter) [inline ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java]: Probably worth checking the exception here as well.

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java (hunk 1 above)
Comment (mbaxter) [reply ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java]: Looks like this test is checking for stalling?  Worth adding the tests for the 2 other cases where it doesn't stall though (T/F and F/T for timeLimitHit/failedRequestsLimitHit).

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java [hunk 3]
```diff
@@ -131,6 +136,18 @@ public void shouldResetRequestsSinceProgressCountWhenProgressIsMade() {
...
+  }
+
   @Test
   public void shouldNotAddRequestsAfterDownloadIsStalled() {
```
Comment (mbaxter) [inline ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java]:
```suggestion
  public void shouldNotAddRequestsAfterDownloadIsCompleted() {
```

Comment (ajsutton) [reply ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java]: 👍

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java]: The test hits the maximum requests without progress limit and checks the download isn't considered stalled until the minimum time condition is also met.  Tidied up the test name and implementation to make this clearer.

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java (hunk 2 above)
Comment (ajsutton) [reply ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java]: 👍

File: ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java (hunk 1 above)
Comment (ajsutton) [reply ethereum/eth/src/test/java/tech/pegasys/pantheon/ethereum/eth/sync/worldstate/WorldDownloadStateTest.java]: Ensured the other combinations of time/requests reached/not-reached were covered.

**Pull request merged**